### DISCUSSION TITLE
Many duplicate captures when querying nodes in file tree

### DISCUSSION THREAD

**Discussion by Symbolk** [tree-sitter/tree-sitter#2101] [Q&A] [answered]
I am not sure this is an expected feature or a bug, but I found myself confused about the captures of the following query, which contains lots of duplicates:

The code:
```
        parser = Parser()
        parser.set_language(JAVA_LANGUAGE)
        tree = parser.parse(bytes(file_content, "utf-8"))
        root_node = tree.root_node

        query = JAVA_LANGUAGE.query("""
            (class_declaration
              name: (identifier) @class_name
              body: (class_body
              (block_comment)? @comment
              (method_declaration) @method
              )
            )
        """)
        captures = query.captures(root_node)
``` 

The length of the captures can be 1000+, with lots of duplicate captures for class_name, comment, and method, all in the one single captures list.

**Comment by maxbrunsfeld** [comment, marked as answer]
That's expected, because that pattern will match once for every method in a class. For every match, it will capture the class name. Also, siblings don't need to be immediately sequential unless you use a `.` (anchor) in  between, so it will match once for every ordered (comment, method) pair. You may want this:

```
            (class_declaration
              name: (identifier) @class_name
              body: (class_body
              (block_comment)? @comment
              .
              (method_declaration) @method
              )
            )

```

**Comment by Symbolk** [comment]
> That's expected, because that pattern will match once for every method in a class. For every match, it will capture the class name. Also, siblings don't need to be immediately sequential unless you use a `.` (anchor) in between, so it will match once for every ordered (comment, method) pair. You may want this:
> 
> ```
>             (class_declaration
>               name: (identifier) @class_name
>               body: (class_body
>               (block_comment)? @comment
>               .
>               (method_declaration) @method
>               )
>             )
> ```

Thanks for the reply! I know this and have tried this (previously I learned this from you in another issue), but this results in another extreme: one single capture for the method. I will try to paste screenshots later~
